Add commands for listing and querying builds

omacranger opened the pull request:
## Description
<!--
Describe the big picture of your changes here to communicate to the maintainers
why we should accept this pull request. If it fixes a bug or resolves a feature
request, be sure to link to that issue.
-->

This PR adds the following commands: 

`saucectl builds list`

```
Returns the list of builds

Usage:
  saucectl builds list [flags]

Aliases:
  list, ls

Flags:
  -h, --help            help for list
  -o, --out string      Output format to the console. Options: text, json. (default "text")
  -p, --page int        Page for pagination. Default is 0.
  -s, --size int        Per page for pagination. Default is 20. (default 20)
      --source string   Job source from saucelabs. Options: vdc, rdc.
      --status string   Filter job using status. Options: passed, failed, error, complete, in progress, queued.
```

`saucectl builds get`

This command by default will find a build by build ID, but can optionally use the `--job-id` flag to search for a build by a job ID instead.

Examples: 

Searches for build by build ID 1234
`saucectl builds get 1234`

Searches for build by job ID 4567
`saucectl builds get --job-id 4567`

```
Usage:
  saucectl builds get [flags]

Flags:
  -h, --help            help for get
      --job-id          Find the build by providing a job ID instead of a build ID.
  -o, --out string      Output format to the console. Options: text, json. (default "text")
      --source string   Job source from saucelabs. Options: vdc, rdc.
```

perf2711 commented on internal/cmd/builds/get.go:
```diff
@@ -0,0 +1,81 @@
+	flags := cmd.PersistentFlags()
+	flags.BoolVarP(&byJob, "job-id", "", false, "Find the build by providing a job ID instead of a build ID.")
+	flags.StringVarP(&out, "out", "o", "text", "Output format to the console. Options: text, json.")
+	flags.StringVar(&buildSource, "source", "", "Build source from saucelabs. Options: vdc, rdc.")
```
In `list.go`, you add `"Required."` to the end of the message. Is this required as well? Do you think it would be a good idea to move common options to a separate file for consistency?

omacranger replied:
Adding the required notice here may be beneficial. I'll see what moving these to a common export would look like.

omacranger replied:
I think this is resolved by nature of moving them to required positional arguments

perf2711 commented on internal/cmd/builds/list.go:
```diff
@@ -0,0 +1,200 @@
+	return nil
+}
+
+func renderListTable(builds []build.Build) {
```
This seems to be used in `get.go`. Do you think it would be a good idea to move this and `defaultTableStyle` to a separate file?

omacranger replied:
This could also probably go into a `shared` / `common` file along with the arg processing. Long term I think it would be good to dedupe all of the `defaulTableStyle` usages across the codebase

omacranger commented on internal/cmd/builds/get.go:
```diff
@@ -0,0 +1,81 @@
+
+			src := build.Source(buildSource)
+			if src != build.SourceRDC && src != build.SourceVDC {
+				return errors.New("invalid build resource. Options: vdc, rdc")
```
This is consistent with other messages in the CLI, but validates and offers one of the two options avialable for source.

omacranger commented on internal/cmd/builds/get.go:
```diff
@@ -0,0 +1,81 @@
+		},
+	}
+	flags := cmd.PersistentFlags()
+	flags.BoolVarP(&byJob, "job-id", "", false, "Find the build by providing a job ID instead of a build ID.")
```
That makes it a bit weird and require explicit validation if you were to include both, no?

perf2711 replied:
I guess it does.

mhan83 commented on internal/cmd/builds/get.go:
```diff
@@ -0,0 +1,81 @@
+	var buildSource string
+
+	cmd := &cobra.Command{
+		Use:          "get",
```
Required positional arg should be included in the usage:

```suggestion
		Use:          "get <buildID>",
```

omacranger replied:
I think this is a good idea. I did go ahead and do the same for `builds list` though. Thoughts?

mhan83 commented on internal/cmd/builds/list.go:
```diff
@@ -0,0 +1,200 @@
+	flags.IntVarP(&page, "page", "p", 0, "Page for pagination. Default is 0.")
+	flags.IntVarP(&size, "size", "s", 20, "Per page for pagination. Default is 20.")
+	flags.StringVarP(&nameFilter, "name", "n", "", "Filter builds by name. Must match full build name.")
+	flags.StringVar(&status, "status", "", "Filter builds using status. Options: passed, failed, error, complete, in progress, queued.")
```
```suggestion
	flags.StringVar(&status, "status", "", "Filter builds using status. Options: success, failed, error, complete, running, queued.")
```

mhan83 reviewed (commented):
If source is required by the builds api, then I think just using the same args as the `jobs` command isn't quite the right fit here. We've generally put required arguments (rightly or wrongly) as positional args, so there's some value here to sticking with that convention.

Thoughts on this alternative:

`saucectl builds list <vdc|rdc>`
`saucectl builds get <buildID>` but if there's no `--source`, just do the user a solid here and look for both rdc and vdc builds for them.

omacranger commented:
> `saucectl builds list <vdc|rdc>` `saucectl builds get <buildID>` but if there's no `--source`, just do the user a solid here and look for both rdc and vdc builds for them.

@mhan83  Do we have any other examples in saucectl where we're doing this for required values? Feels odd. The API docs make it pretty explicit that those are stored in two separate locations so doing two queries and discarding one error, or only surfacing them if two occur, feels strange.

omacranger commented on internal/build/build.go:
```diff
@@ -6,14 +6,55 @@ import (
+	SourceRDC Source = "rdc" // Real Device Cloud
+)
+
+type Status string
+
+const (
+	StateRunning  Status = "running"
+	StateError    Status = "error"
+	StateFailed   Status = "failed"
+	StateComplete Status = "complete"
+	StateSuccess  Status = "success"
+)
+
+var AllStates = []Status{StateRunning, StateError, StateFailed, StateComplete, StateSuccess}
```
The field in the API is called `status`, so I think it makes more sense to keep that name -- but I will unify that for all the variables here.

mhan83 commented:
> > `saucectl builds list <vdc|rdc>` `saucectl builds get <buildID>` but if there's no `--source`, just do the user a solid here and look for both rdc and vdc builds for them.
> 
> @mhan83 Do we have any other examples in saucectl where we're doing this for required values? Feels odd. The API docs make it pretty explicit that those are stored in two separate locations so doing two queries and discarding one error, or only surfacing them if two occur, feels strange.

Yup, the `saucectl artifacts list <jobID>` command works like this. RDC and VDC job artifacts are handled by different end points, but the command figures it out by querying.

If it feels strange, that's fine; it was just the first thing that popped into my head. I still don't think a required arg should be a flag, but this was simply a subjective preference by the devx team and its fine to deviate from that.

omacranger commented:
> Yup, the `saucectl artifacts list <jobID>` command works like this. RDC and VDC job artifacts are handled by different end points, but the command figures it out by querying.
> 
> If it feels strange, that's fine; it was just the first thing that popped into my head. I still don't think a required arg should be a flag, but this was simply a subjective preference by the devx team and its fine to deviate from that.

I agree with the required arg being a flag, I've refactored that into a positional arg where required and added validation for both. I also moved some of the table definitions out into a shared / common file to reduce the duplicates.

mhan83 reviewed (commented):
Something still feels a bit off...

`saucectl builds get <vdc|rdc> <buildID>`

But `buildID` isn't really a required argument. It could be omitted in place of `--job-id`.

And now I'm wondering, if you're adding a `get` subcommand, could a user reasonably expect to get the jobs belonging to the identified build? I kind of think that it is a reasonable expectation which makes me think `get` isn't the right fit here.

I honestly still prefer the suggested interface in the ticket: treat fetching a build by jobID as a kind of filter, even if that's not how its modelled by the API.
